File: UI/src/lib/api/analytics.ts
import { api } from './client';

// Types

export interface AnalyticsOverview {
	total_issues: number;
	open_issues: number;
	completed_issues: number;
	overdue_issues: number;
	total_projects: number;
	total_members: number;
	started_issues?: number;
	unassigned_issues?: number;
	completion_rate?: number;
	avg_lead_time_hours?: number | null;
	avg_cycle_time_hours?: number | null;
}

export interface DistributionItem {
	status_id: string;
	name: string;
	color: string | null;
	category: string;
	count: number;
}

export interface PriorityDistributionItem {
	priority: number;
	count: number;
}

export interface AnalyticsDistribution {
	by_status?: DistributionItem[];
	by_priority?: PriorityDistributionItem[];
}

export type AnalyticsMeasure = 'issue_count' | 'issue_age' | 'lead_time' | 'cycle_time' | 'triage_time';
export type AnalyticsSlice =
	| 'none'
	| 'status_type'
	| 'status'
	| 'priority'
	| 'assignee'
	| 'team'
	| 'project'
	| 'cycle'
	| 'label'
	| 'creator';
export type AnalyticsSegment = AnalyticsSlice;

export interface AnalyticsFilterParams {
	team_id?: string;
	project_id?: string;
	cycle_id?: string;
	assignee_id?: string;
	creator_id?: string;
	status_id?: string;
	status_type?: string;
	priority?: string;
	label_id?: string;
	include_sub_issues?: boolean;
	include_triage?: boolean;
}

export interface AnalyticsScopeParams {
	team_id?: string;
}

export interface InsightsParams extends AnalyticsFilterParams {
	measure: AnalyticsMeasure;
	slice?: AnalyticsSlice;
	segment?: AnalyticsSegment;
	from?: string;
	to?: string;
}

export interface InsightsGroup {
	key: string;
	label: string;
	color?: string | null;
	count?: number;
	value?: number | null;
	p50?: number | null;
	p75?: number | null;
	p95?: number | null;
	segments?: InsightsGroup[];
}

export interface InsightsPoint {
	issue_id: string;
	identifier: string;
	title: string;
	value?: number | null;
	slice_key?: string;
	segment_key?: string;
}

export interface AnalyticsInsights {
	measure: AnalyticsMeasure;
	slice?: AnalyticsSlice;
	segment?: AnalyticsSegment;
	unit?: string;
	total_count?: number;
	aggregate?: number | null;
	groups?: InsightsGroup[];
	points?: InsightsPoint[];
}

export interface BurnupParams extends AnalyticsFilterParams {
	from: string;
	to: string;
	interval?: 'day' | 'week' | 'month';
}

export interface BurnupPoint {
	date: string;
	created?: number;
	completed?: number;
	total_created?: number;
	total_completed?: number;
	scope?: number;
}

export interface AnalyticsBurnup {
	interval: string;
	from: string;
	to: string;
	points?: BurnupPoint[];
}

// Helpers

function dateString(d: Date): string {
	const y = d.getFullYear();
	const m = String(d.getMonth() + 1).padStart(2, '0');
	const day = String(d.getDate()).padStart(2, '0');
	return `${y}-${m}-${day}`;
}

export function defaultDateRange(dayCount = 90): { from: string; to: string } {
	const today = new Date();
	today.setHours(12, 0, 0, 0);
	const to = dateString(today);
	const from = new Date(today);
	from.setDate(from.getDate() - Math.max(0, dayCount - 1));
	return { from: dateString(from), to };
}

function buildQuery(params: AnalyticsScopeParams | InsightsParams | BurnupParams): string {
	const query = new URLSearchParams();
	for (const [key, value] of Object.entries(params)) {
		if (typeof value === 'boolean') query.set(key, value ? 'true' : 'false');
		if (typeof value === 'string' && value !== '' && value !== 'none') query.set(key, value);
	}
	const qs = query.toString();
	return qs ? `?${qs}` : '';
}

// API functions

export function getAnalyticsOverview(slug: string, params: AnalyticsScopeParams = {}): Promise<AnalyticsOverview> {
	return api.get<AnalyticsOverview>(`/api/workspaces/${slug}/analytics/overview${buildQuery(params)}`);
}

export function getAnalyticsDistribution(
	slug: string,
	params: AnalyticsScopeParams = {}
): Promise<AnalyticsDistribution> {
	return api.get<AnalyticsDistribution>(`/api/workspaces/${slug}/analytics/distribution${buildQuery(params)}`);
}

export function getAnalyticsInsights(slug: string, params: InsightsParams): Promise<AnalyticsInsights> {
	const query = buildQuery(params);
	return api.get<AnalyticsInsights>(`/api/workspaces/${slug}/analytics/insights${query}`);
}

export function getAnalyticsBurnup(slug: string, params: BurnupParams): Promise<AnalyticsBurnup> {
	const query = buildQuery(params);
	return api.get<AnalyticsBurnup>(`/api/workspaces/${slug}/analytics/burnup${query}`);
}

File: BE/internal/dto/analytics.go
package dto

type AnalyticsScopeParams struct {
	TeamID string `query:"team_id"`
}

type AnalyticsOverview struct {
	TotalIssues       int     `json:"total_issues"`
	OpenIssues        int     `json:"open_issues"`
	CompletedIssues   int     `json:"completed_issues"`
	OverdueIssues     int     `json:"overdue_issues"`
	TotalProjects     int     `json:"total_projects"`
	TotalMembers      int     `json:"total_members"`
	StartedIssues     int     `json:"started_issues"`
	UnassignedIssues  int     `json:"unassigned_issues"`
	CompletionRate    float64 `json:"completion_rate"`
	AvgLeadTimeHours  float64 `json:"avg_lead_time_hours"`
	AvgCycleTimeHours float64 `json:"avg_cycle_time_hours"`
}

type StatusCount struct {
	StatusID   string  `json:"status_id" db:"status_id"`
	StatusName string  `json:"name" db:"name"`
	Category   string  `json:"category" db:"category"`
	Color      *string `json:"color" db:"color"`
	Count      int     `json:"count" db:"count"`
}

type PriorityCount struct {
	Priority int `json:"priority" db:"priority"`
	Count    int `json:"count" db:"count"`
}

type AssigneeCount struct {
	AssigneeID *string `json:"assignee_id" db:"assignee_id"`
	Count      int     `json:"count" db:"count"`
}

type AnalyticsIssueDistribution struct {
	ByStatus   []StatusCount   `json:"by_status"`
	ByPriority []PriorityCount `json:"by_priority"`
}

type AnalyticsInsightsParams struct {
	Measure          string  `query:"measure"`
	Slice            string  `query:"slice"`
	Segment          string  `query:"segment"`
	From             string  `query:"from"`
	To               string  `query:"to"`
	TeamID           *string `query:"team_id"`
	ProjectID        *string `query:"project_id"`
	CycleID          *string `query:"cycle_id"`
	AssigneeID       *string `query:"assignee_id"`
	CreatorID        *string `query:"creator_id"`
	StatusID         *string `query:"status_id"`
	StatusType       *string `query:"status_type"`
	Priority         *string `query:"priority"`
	LabelID          *string `query:"label_id"`
	IncludeSubIssues *bool   `query:"include_sub_issues"`
	IncludeTriage    *bool   `query:"include_triage"`
}

type AnalyticsGroup struct {
	Key       string             `json:"key"`
	Label     string             `json:"label"`
	Color     string             `json:"color"`
	Count     int                `json:"count"`
	Value     float64            `json:"value"`
	Aggregate *float64           `json:"aggregate,omitempty"`
	P50       *float64           `json:"p50,omitempty"`
	P75       *float64           `json:"p75,omitempty"`
	P95       *float64           `json:"p95,omitempty"`
	Segments  []AnalyticsSegment `json:"segments"`
}

type AnalyticsSegment struct {
	Key       string   `json:"key"`
	Label     string   `json:"label"`
	Color     string   `json:"color"`
	Count     int      `json:"count"`
	Value     float64  `json:"value"`
	Aggregate *float64 `json:"aggregate,omitempty"`
	P50       *float64 `json:"p50,omitempty"`
	P75       *float64 `json:"p75,omitempty"`
	P95       *float64 `json:"p95,omitempty"`
}

type AnalyticsPoint struct {
	IssueID    string  `json:"issue_id"`
	Identifier string  `json:"identifier"`
	Title      string  `json:"title"`
	Value      float64 `json:"value"`
	SliceKey   string  `json:"slice_key"`
	SegmentKey string  `json:"segment_key"`
}

type AnalyticsInsightsResponse struct {
	Measure    string           `json:"measure"`
	Slice      string           `json:"slice"`
	Segment    string           `json:"segment"`
	Unit       string           `json:"unit"`
	TotalCount int              `json:"total_count"`
	Aggregate  float64          `json:"aggregate"`
	Groups     []AnalyticsGroup `json:"groups"`
	Points     []AnalyticsPoint `json:"points"`
}

type AnalyticsBurnupParams struct {
	From             string  `query:"from"`
	To               string  `query:"to"`
	Interval         string  `query:"interval"`
	TeamID           *string `query:"team_id"`
	ProjectID        *string `query:"project_id"`
	CycleID          *string `query:"cycle_id"`
	AssigneeID       *string `query:"assignee_id"`
	StatusType       *string `query:"status_type"`
	IncludeSubIssues *bool   `query:"include_sub_issues"`
	IncludeTriage    *bool   `query:"include_triage"`
}

type AnalyticsBurnupPoint struct {
	Date           string `json:"date"`
	Created        int    `json:"created"`
	Completed      int    `json:"completed"`
	TotalCreated   int    `json:"total_created"`
	TotalCompleted int    `json:"total_completed"`
	Scope          int    `json:"scope"`
}

type AnalyticsBurnupResponse struct {
	Interval string                 `json:"interval"`
	From     string                 `json:"from"`
	To       string                 `json:"to"`
	Points   []AnalyticsBurnupPoint `json:"points"`
}
